### ISSUE TITLE
Hercules and blinkenlights: accessing Hercules hardware state by external program

### ISSUE THREAD

**Issue by macieksoft** [SDL-Hercules-390/hyperion#376]
I see there are some GUI programs for Hercules (SDL HercGUI, Jason and so on). How do they interface with the Hercules emulator? (i.e. how does data about registers, IO channels, IO devices etc, get _from_ Hercules _to_ the external application? Is there some kind of SDK to do that?)

And same thing the other way around too: how is data about various actions (IPL, RESET, STOP, START, EXT, changing memory content, etc) sent _from_ the external application _to_ Hercules?

**Comment by Fish-Git** [comment]
Hi (whoever you are)! &nbsp; **`(*)`**

External GUIs and Hercules currently communicate with each other via a simple stdio pipes.

Whenever Hercules is started with the `-e` or `--externalgui` option (or, alternatively, with the string "EXTERNALGUI" specified as the very last option on the command line), Hercules sets an internal flag (called `extgui`) to true and then loads the "dyngui" module that then controls all communication between itself (i.e. Hercules) and the other end of the pipe (i.e. the external GUI).

As `dyngui` (the Hercules module that controls communication between itself and the external GUI) runs, it periodically queries the status of certain internal values, such as the contents of registers, the status of devices, etc) and "sends" that status information to the external GUI via a simple text string written to the stderr pipe which the external GUI -- running on the same system that Hercules is running on -- then reads from its end of that same pipe and uses to update its "blinkenlights".

Hercules's log messages (messages that it normally displays on the screen) are written to the stdout pipe instead whenever the `extgui` flag is enabled, which the external GUI at the other end of the pipe then reads and displays in the external GUI's window.

Commands are sent from the external GUI to Hercules via the stdin pipe. The external GUI writes the command to its end of the stdin pipe, and Hercules's `dyngui` module then reads the command from its end of the same pipe.

Normal Hercules messages (written to the stdout pipe) are formatted normally, exactly how you would normally see them on the screen.

"Status" messages however (written to the stderr pipe) are specially formatted to identify what type of status information is being sent. 32-bit register value status messages for example, are prefixed with the string "GR0-3=", etc. The external GUI then, upon receipt of such a status message at its end of stderr pipe, simply parses this textual information and uses it to update its blinkenlights.

There are also some other specially formatted stdin commands the external GUI can use to request the enabling or disabling of certain options such as whether or not to start or stop sending register status updates **`(**)`** which use the format "]xxxx=zzz" (where xxx is the name of the option and zzz is usually 1 or 0 to enable or disable the option). If the command that is sent is not "]xxx=yyy" however, then Hercules's `dyngui` treats it as a Hercules command.

None of this is documented anywhere.

My suggestion would be to study closely how Hercules's "dyngui.c" module behaves for more insight into the how it all works.

Also note that all of Hercules's stand-alone utilities also support handshaking with an external GUI in exactly the same way: via stdio pipes. This allows the external GUI to start a Hercules utility (such as dasdinit for example) from its own user interface and receive the utility's "status messages" via the stderr pipe and normal screen messages via the stdout pipe. (stdin is never used for utilities) The "status" messages that utilities write to the stderr pipe are usually just simple "progress" type messages, such as a "percentage complete". The external GUI then uses this to update its dialog's progress bar.

That's pretty much it. &nbsp; _&lt;shrug&gt;_

Hope that helps!

Unless you have some other questions, I am going to go ahead and close this Issue. If you need to ask additional questions or need additional clarification regarding anything I've written, feel free to re-open this issue and I will try to answer to the best of my ability.

Thanks. &nbsp; _(whoever you are!)_ &nbsp; `;-)`

_____
**`(*)`** _It would be so much nicer if people would get into the habit of either signing their posts with their name or else identifying themselves via their GitHub account information so that we would know who it was we were conversing with!_

_While I certainly understand some people may indeed wish to remain anonymous, in my personal experience such is usually not the case. Rather, they simple forget to do so purely out of habit._

_**Bottom line:** It is so much more enjoyable (friendly and cordial) conversing with someone by name than otherwise!_ &nbsp; `:)`

**`(**)`** _An external GUI may wish to do this when, for example, the external GUI user chooses to not show the "Registers" dialog on their screen. Thus, if the external GUI is not displaying register information anywhere on its screen, it makes sense to ask Hercules to not waste time sending it since it will never be used. Similarly, whenever the  external GUI user chooses to open/display the Register Status window on their external GUI window, the external GUI would then ask Hercules to "please begin sending me register status information again!"._
